### PR TITLE
WIP: Refactoring resource logic DAG

### PR DESCRIPTION
**Pull request by degregat**
@isheff can you check if the section on transactions in resources.md is compatible with what typhon requires?
(the section on the controllers is not very comprehensive yet, so maybe we should have a more extensive discussion and update that subsection before you review it.)

paging @cwgoes for general review

### CODE REVIEW

**Comment by isheff** [comment]
What does ["this way" refer to here](https://github.com/anoma/specs/blob/degregat/resource-dag-refactor/src/architecture/layers/resources.md#resources)? It sounds like the "type" can be derived from "a partially ordered history of the immutable resources" provided by a "suffix," but this seems contradictory to [the previous section, wherein "The Type of a Resource and its fungibility, is determined by the Resource Logic, Static Resource Data and its Prefix](https://github.com/anoma/specs/blob/degregat/resource-dag-refactor/src/architecture/layers/resources.md#resource-types)."

**Comment by isheff** [comment]
I'm confused about what the [Components section](https://github.com/anoma/specs/blob/degregat/resource-dag-refactor/src/architecture/layers/resources.md#components) refers to. At first I thought it was a list of the components that make up each resource, but then I saw that [Resources are a component](https://github.com/anoma/specs/blob/degregat/resource-dag-refactor/src/architecture/layers/resources.md#resources). I guess they're supposed to be components of the resource management layer? 

It would really help to have a clear list of components of a resource somewhere, especially _before_ we start [referring to them](https://github.com/anoma/specs/blob/degregat/resource-dag-refactor/src/architecture/layers/resources.md#resource-types).

**Comment by isheff** [comment]
The statement "[Suffix, providing a partially ordered history of the immutable Resources that inhabited a type can be derived this way](https://github.com/anoma/specs/blob/degregat/resource-dag-refactor/src/architecture/layers/resources.md#resources). " is not an adequate definition of what a Suffix is, or at least I hope not. This makes it sound like a Suffix is a DAG of potentially infinite size. That sounds unwieldy. 

Also, the word "Suffix" means "a [morpheme](https://www.google.com/search?q=morpheme&si=AMnBZoFm76bvId4K9j6r5bU9rVYrk9jc6IElGMO1hQ27h29T1EG0IDlU8XiUKp6zxDhhQyNdz2zcTcZq6gODFSiRFYwMeIDlyw%3D%3D&expnd=1) added at the end of a word to form a derivative," and it's not at all clear from this definition by what leap of intuition this thing could be called a "suffix."

**Comment by isheff** [comment]
"Prefix" is never defined, except perhaps as "[encoded as a List of ContentHashes](https://github.com/anoma/specs/blob/degregat/resource-dag-refactor/src/architecture/layers/resources.md#resource_data_static)"

**Comment by isheff** [comment]
Given that "The Type of a Resource and its fungibility, is determined by the Resource Logic, Static Resource Data and its Prefix," and that static resource data contains Prefix, are we to understand that "The Type of a Resource" contains "Prefix" twice?

**Comment by isheff** [comment]
[I thought we agreed not to give the word "key" any more definitions. Perhaps "resource hash."](https://github.com/anoma/specs/blob/degregat/resource-dag-refactor/src/architecture/layers/resources.md#resource-key)

**Comment by isheff** [comment]
We're going to need an example of where [dynamic predicates](https://github.com/anoma/specs/blob/degregat/resource-dag-refactor/src/architecture/layers/resources.md#predicates) are used, because from this definition, it seem slike literally any encoding of any conceivable transaction constraint is a dynamic predicate, which doesn't seem like a helpful thing to define.

**Comment by isheff** [comment]
We say that [some predicates encode Resource logics](https://github.com/anoma/specs/blob/degregat/resource-dag-refactor/src/architecture/layers/resources.md#predicates), and all other predicates are "dynamic predicates." What makes them "dynamic"? Also, what do we call "predicates that encode resource logics"?

We then say [The Resource Logic of a Resource is defined via a static Predicate](https://github.com/anoma/specs/blob/degregat/resource-dag-refactor/src/architecture/layers/resources.md#resource-logic). This raises several questions:
- does each resource have exactly 1 resource logic? If so, this should go in the list of components of a resource. Also, we should have a list of components of a resource. 
- What is a "static Predicate"? This was never defined. What makes it "static"?
- What is the difference between a "static Predicate" and a "resource logic"? Why have 2 words for what appear to be the same thing?

**Comment by isheff** [comment]
What is an "address"? [Concerning Resource Logics, we say "Its address is computed via hash(Predicate)](https://github.com/anoma/specs/blob/degregat/resource-dag-refactor/src/architecture/layers/resources.md#resource-logic)" We later refer to "[The Value of the encoded Resource represented by a ByteString (i.e. the current content of the Memory behind the address).](https://github.com/anoma/specs/blob/degregat/resource-dag-refactor/src/architecture/layers/resources.md#resource_data_dynamic)." I thought resources were themselves immutable, which would make sense for the way the address is used in reference to Predicates, but not as its used in reference to "Value"s.

**Comment by isheff** [comment]
I'm not sure how to interpret the phrase "[The Resource Logic specifies under what conditions Resources of that type...](https://github.com/anoma/specs/blob/degregat/resource-dag-refactor/src/architecture/layers/resources.md#resource-logic)" We had previously said that Resource Logics were attached to specific Resources, not to specific Types. 

What is the difference between a Type and a Resource Logic? If there is a difference, perhaps each resource should reference a Type, and each Type should reference a Resource Logic?

**Comment by isheff** [comment]
What does the note "[The only semantic differentiation between Predicates relevant to Applications is between ones that are static and influence fungibility and ones that are dynamic and don't influence fungibility](https://github.com/anoma/specs/blob/degregat/resource-dag-refactor/src/architecture/layers/resources.md#resource-logic)" mean? Surely applications care about lots of semantics other than whether or not it's fungible.

**Comment by isheff** [comment]
Does a [shielded ptx](https://github.com/anoma/specs/blob/degregat/resource-dag-refactor/src/architecture/layers/resources.md#partial-transaction-ptx) fundamentally need to have the same number of input and output resources? Why must the number of resources be conserved?

**Comment by isheff** [comment]
What is the[ executable in a PartialTx](https://github.com/anoma/specs/blob/degregat/resource-dag-refactor/src/architecture/layers/resources.md#partial-transaction-ptx) for?

**Comment by isheff** [comment]
As written, we call the same `logic` function for resources that are being consumed as those being created. Shouldn't these be different? Perhaps 
`(all (map (\r consumeLogic r inr outr) inr)) and (all (map (\r createLogic r inr outr) outr))`

**Comment by isheff** [comment]
As written, each `ptx` in a valid transaction must be individually valid. Since `ptx`s are of limited size, this means you can't do anything that would require atomically, say, splitting a resource into 3 parts. Is this intentional? Is it a limitation of our system or just a writing mistake? What if we only wanted to check validity for the Transaction as a whole?

**Comment by isheff** [comment]
What is the motivation behind "[A transaction is balanced if and only if the input and output sets of each Resource Type are of the same size.](https://github.com/anoma/specs/blob/degregat/resource-dag-refactor/src/architecture/layers/resources.md#transactions-tx)" I still don't understand this idea of "balance" at all. Why would we want to ensure we create the same number of resources of each type as we consume? Couldn't we, for instance, start with a single UTXO representing 2 money, and end up with 2 UTXOs representing 1 money each? 

Or, for definitions of "size" that take simple fungibility into account, what about invariants that can't be expressed as simple fungibility? Assuming we have some mechanism to deal with those, why can't we use that more general mechanism to encode fungibility? Why is there a special case for this in the spec?

**Comment by isheff** [comment]
It seems like the definition "[A transaction is valid if and only if the ptxs it contains are valid and it is balanced](https://github.com/anoma/specs/blob/degregat/resource-dag-refactor/src/architecture/layers/resources.md#transactions-tx)." is missing something. By this definition, any transaction could be broken up into all of its ptxs, and each of those alone would be a valid transaction. What then is the point of the larger "transaction"?

**Comment by isheff** [comment]
What is "the given Resource" in the first sentence of the "[Controllers](https://github.com/anoma/specs/blob/degregat/resource-dag-refactor/src/architecture/layers/resources.md#controllers)" section? Who gives it?

**Comment by isheff** [comment]
In the [Controllers](https://github.com/anoma/specs/blob/degregat/resource-dag-refactor/src/architecture/layers/resources.md#controllers) section, what does the phrase "The first controller in the list" mean? This is the first we've heard of a list of any kind. Which list are we talking about? I think we should have somewhere a list of all the components of a resource.

**Comment by isheff** [comment]
After reading the[ Transaction Execution Logic](https://github.com/anoma/specs/blob/degregat/resource-dag-refactor/src/architecture/layers/resources.md#transaction-execution-logic) section, I still don't know what this is. Is this a collection of executables? Is this collection a field of a Transaction? A Type? A Resource Logic? A Controller (somehow)? 

This sounds a lot like a field we discussed for a Transaction _Candidate_, but it's not the same, since that field is mandatory, and must be deterministic, so it is definitely not "up to the computing parties whether they use the Executable."

**Comment by isheff** [comment]
Why does a `Resource` contain a `ResourceLogic` rather than `address` of a `ResourceLogic`?

**Comment by isheff** [comment]
The type of `ResourceLogic.predicate` does not look anything like the function call `logic r inr outr` used earlier. Are these supposed to represent related things?

**Comment by isheff** [comment]
I do not understand the point of `resource_data_static.` Why not encode all of this stuff directly in the predicate in `ResourceLogic`? That seems strictly equivalent.

**Comment by isheff** [comment]
By the time I read [`resource_data_dynamic`](https://github.com/anoma/specs/blob/degregat/resource-dag-refactor/src/architecture/layers/resources.md#resource_data_dynamic), I still have no idea what "relevant to the fungibility of a _Resource_" means. Is the idea here that nothing in `resource_data_dynamic` should be readable by predicates determining whether this Resource can be created or destroyed? What's the point of this field then?

**Comment by isheff** [comment]
I'm not sure how to interpret "The dynamic Suffix identifying the version of the Value of the Resource. Should always be a nonce." Resources have Versions and Values now? Versions and Values were never mentioned elsewhere in the spec.

**Comment by isheff** [comment]
Why would I put a Dynamic predicate in a resource? Surely I would encode that in the resource's predicate directly.

**Comment by isheff** [comment]
Why would I ever [encode an unbalanced `ptx`](https://github.com/anoma/specs/blob/degregat/resource-dag-refactor/src/architecture/layers/resources.md#intent)? You can't put them in a transaction, and it's not like you can make a new balanced `ptx` out of multiple unbalanced `ptx` (or if you can, this spec says nothing about it).

**Comment by isheff** [comment]
The first time we hear about nullifier keys is in the [Ownership](https://github.com/anoma/specs/blob/degregat/resource-dag-refactor/src/architecture/layers/resources.md#ownership) section. What is a nullifier key? Is this not something that's encoded in the resource logic? Why is there another thing in addition to the resource logic?

**Comment by isheff** [comment]
What does it mean for a set of [partial candidate transactions to be balanced](https://github.com/anoma/specs/blob/degregat/resource-dag-refactor/src/architecture/layers/resources.md#candidate-transaction)? Transaction candidates can dynamically generate output resources after ordering time. You can't know in advance if they're balanced.

**Comment by isheff** [comment]
I have no idea what's going on in the [Optional Execution](https://github.com/anoma/specs/blob/degregat/resource-dag-refactor/src/architecture/layers/resources.md#optional-execution-optionally-using-executable-from-the-tel) step. This sounds vaguely like but is not at all compatible with what we're hoping to do in Typhon.

**Comment by isheff** [comment]
The first time we hear about Resource and Nullifier trees is in [Apply Transaction](https://github.com/anoma/specs/blob/degregat/resource-dag-refactor/src/architecture/layers/resources.md#apply-transaction). What are those, according to this spec?

**Comment by isheff** [comment]
From what I can tell, `resources.md` is not compatible with what Typhon requires. @degregat and I should schedule a meeting.

**Comment by cwgoes** [comment]
Hi! Just chiming in on a few of these questions in an attempt to clarify a bit. I think a synchronous discussion would also help.

> In the [Controllers](https://github.com/anoma/specs/blob/degregat/resource-dag-refactor/src/architecture/layers/resources.md#controllers) section, what does the phrase "The first controller in the list" mean? This is the first we've heard of a list of any kind. Which list are we talking about? I think we should have somewhere a list of all the components of a resource.

See https://hackmd.io/@heliax/ByOall6mn for an explanation of controllers (which should be merged into this document, if we agree with it).

> As written, [we call the same logic function for resources that are being consumed as those being created.](https://github.com/anoma/specs/blob/degregat/resource-dag-refactor/src/architecture/layers/resources.md#resource-logic) Shouldn't these be different? Perhaps (all (map (\r consumeLogic r inr outr) inr)) and (all (map (\r createLogic r inr outr) outr))

This distinction can easily be made within the resource logic itself (although we should pass the logic a bit to indicate whether the resource is being created or consumed, which should be made clear in this function).

> Or, for definitions of "size" that take simple fungibility into account, what about invariants that can't be expressed as simple fungibility? Assuming we have some mechanism to deal with those, why can't we use that more general mechanism to encode fungibility? Why is there a special case for this in the spec?

I broadly agree that there doesn't need to be a special case, at some layer of abstraction. At some layer of abstraction (currently the resource logic layer), having this concept of "balance" is helpful, because:
- It composes (if A is balanced, and B is balanced, A <> B is balanced)
- It's easy to implement the check in a way which is additively homomorphic (as Taiga does)
- It's possible to encode any more complex concept of "balance" in terms of compositions of this simple concept (I think)

> I had imagined something a little more like this: The Transaction Candidate features a mandatory executable, which reads whatever values Typhon can give it after ordering time, and then outputs some new resources. In the case of a fully shielded transaction, this executable is likely to be very simple: it just outputs the resources. We'd just be using the executable as a way to encode a list. Anway, Taiga then checks the output resources for validity and balance and whatnot, and only if it succeeds do we store those resources output.

Yes, we need both of these things - we're using the word "executable" to refer to distinct concepts at different levels of abstraction, which I think is probably part of the confusion. There always needs to be a transaction executable, in the sense in which you refer to it. However, we also want a separation of concerns within the transaction execution logic such that, after being called by Typhon:
(a) An executable within the transaction can be run, potentially reading resources and outputting new ones
(b) All resource logics associated with consumed and created resources are checked along with balance

This separation is important because it means that the executable in (a) doesn't need to be trusted (other than the transaction submitter paying for enough gas or whatever), since all of the correctness invariants are checked in (b).

**Comment by degregat** [comment]
> What does the note "[The only semantic differentiation between Predicates relevant to Applications is between ones that are static and influence fungibility and ones that are dynamic and don't influence fungibility](https://github.com/anoma/specs/blob/degregat/resource-dag-refactor/src/architecture/layers/resources.md#resource-logic)" mean? Surely applications care about lots of semantics other than whether or not it's fungible.

Applications are allowed to care about different things, but these concerns should be mapped to a combination of Ephemeral Resources + Resource Logics and Dynamic Predicates on the level of the Resource Management System.

I'll change it to: "The only semantic differentiation _relevant for the resource management system_ is ..."

Does that sound good?

**Comment by degregat** [comment]
> Does a [shielded ptx](https://github.com/anoma/specs/blob/degregat/resource-dag-refactor/src/architecture/layers/resources.md#partial-transaction-ptx) fundamentally need to have the same number of input and output resources? 

I assume because of Taiga Implementation details, which I assume stem from the properties of the cryptography.

> Why must the number of resources be conserved?

Same reason as above.

@joebebel @vveiln is that correct?

**Comment by degregat** [comment]
> As written, [we call the same `logic` function for resources that are being consumed as those being created.](https://github.com/anoma/specs/blob/degregat/resource-dag-refactor/src/architecture/layers/resources.md#resource-logic) Shouldn't these be different? Perhaps `(all (map (\r consumeLogic r inr outr) inr)) and (all (map (\r createLogic r inr outr) outr))`

All predicates of all resources within a ptx share the same scope/evaluation context, namely the collection of resources (with visibility of all their fields) in that ptx, since the creation constraints might refer to the consumption constraints (and vice versa) of the same or other resources.

We started out with separate Creation and Consumption Predicates, but after the above became apparent, we  unified them.

**Comment by degregat** [comment]
> As written, each `ptx` in a valid transaction must be individually valid. Since `ptx`s are of limited size, this means you can't do anything that would require atomically, say, splitting a resource into 3 parts. Is this intentional? Is it a limitation of our system or just a writing mistake? 

It is a limitation of the system. We still need to decide which value for _k_ would be sensible, or if we could support multiple different ones (I assume that would lead to a split of anonymity sets though).

> What if we only wanted to check validity for the Transaction as a whole?

We should be able to implement a special case where ptx's don't need to be individually valid in the transparent case.

**Comment by degregat** [comment]
> It seems like the definition "[A transaction is valid if and only if the ptxs it contains are valid and it is balanced](https://github.com/anoma/specs/blob/degregat/resource-dag-refactor/src/architecture/layers/resources.md#transactions-tx)." is missing something. By this definition, the only way to encode a property you want for a transaction is through balance. Why can't I encode some more intricate property?

As far as I understand, more intricate properties would need to be expressed via ptx's.
Again, we should be able to support special cases in the transparent case.

@joebebel @vveiln @Acentelles is this correct?

**Comment by joebebel** [comment]
> Does a [shielded ptx](https://github.com/anoma/specs/blob/degregat/resource-dag-refactor/src/architecture/layers/resources.md?rgh-link-date=2023-05-02T10%3A05%3A55Z#partial-transaction-ptx) fundamentally need to have the same number of input and output resources?

In some sense, this is mandatory, because there is no flexibility in the interface for a private VP, the VP has to see the entire PTX and if we make some PTX bigger, then every PTX incurs the increase in overhead. So it's better to make the fixed size of a PTX as small as reasonable.

On the other hand, functionally, too large PTX are not a problem because we can make "dummy" resources to fill any unused slots.

> As written, each ptx in a valid transaction must be individually valid. Since ptxs are of limited size, this means you can't do anything that would require atomically, say, splitting a resource into 3 parts. Is this intentional? Is it a limitation of our system or just a writing mistake?

No, you can still do atomic operations. Yes, this is something we designed for very intentionally, and no, it's not a limitation.

The primary way you can bind PTX together is using ephemeral state. This is a fully general mechanism and also how we handle intents. E.g. you can make PTX 1 conditional on PTX 2, by creating some ephemeral resource in PTX 1 that must be consumed in PTX 2. Now, that said, we might want to add other ways to make groups of PTX atomic.

> It seems like the definition "[A transaction is valid if and only if the ptxs it contains are valid and it is balanced](https://github.com/anoma/specs/blob/degregat/resource-dag-refactor/src/architecture/layers/resources.md?rgh-link-date=2023-05-02T10%3A23%3A49Z#transactions-tx)." is missing something. By this definition, the only way to encode a property you want for a transaction is through balance. Why can't I encode some more intricate property?

I think this is a matter of perspective, because from the execution engine point of view, these are the _only_ properties it can check. There would be other properties implied by these, which provers have to prove. We could discuss further here the implied properties, such as how the VP of each resource is called.

**Comment by degregat** [comment]
> The type of `ResourceLogic.predicate` does not look anything like the function call `logic r inr outr` used earlier. Are these supposed to represent related things?

As far as I understand it means "take the logic of Resource r, which needs input and output resource sets as arguments". @cwgoes is that correct? 

It seems haskell syntax is no ideal, what would you choose @isheff ?

**Comment by degregat** [comment]
> 



> I do not understand the point of `resource_data_static.` Why not encode all of this stuff directly in the predicate in `ResourceLogic`? That seems strictly equivalent.

See [here](https://github.com/anoma/specs/pull/139#issuecomment-1531203601).

**Comment by joebebel** [comment]
> I do not understand the point of resource_data_static. Why not encode all of this stuff directly in the predicate in ResourceLogic? That seems strictly equivalent.

Sure, this is equivalent in theory, because code=data, but it's not equivalent in practice because code (whether wasm or circuits) is just structured differently than static data and has different requirements. 

For example, we could encode intent predicates entirely inside a circuit (e.g. Alice embeds asset types, amounts, keys, etc into the intent predicate circuit). The problem is that reading this data from the circuit description is nearly impossible. Presumably it's possible, albeit awkward, if you have access to the halo2 crate (and vamp-ir toolchain) but it's definitely impossible to do inside another circuit. So if we want to have another predicate which can inspect an intent resource and extract static data, then this static data has to be segregated from the code.

**Comment by joebebel** [comment]
> We're going to need an example of where [dynamic predicates](https://github.com/anoma/specs/blob/degregat/resource-dag-refactor/src/architecture/layers/resources.md?rgh-link-date=2023-04-29T20%3A59%3A36Z#predicates) are used, because from this definition, it seem slike literally any encoding of any conceivable transaction constraint is a dynamic predicate, which doesn't seem like a helpful thing to define.

In my mind, the prototypical example of a dynamic predicate would be a user's VP. So a token resource can have a resource logic which is the same for all instantiations of this resource, and the token resource logic has a concept of "owner" specified by a predicate in `resource_data_dynamic`, which has to also be satisfied in order to transfer tokens. The "owner" wouldn't go in the resource logic or `resource_data_static` because tokens with the same resource logic but different owners should be fungible. 

> We say that [some predicates encode Resource logics](https://github.com/anoma/specs/blob/degregat/resource-dag-refactor/src/architecture/layers/resources.md?rgh-link-date=2023-04-29T21%3A03%3A28Z#predicates), and all other predicates are "dynamic predicates." What makes them "dynamic"? Also, what do we call "predicates that encode resource logics"?

At least when we were naming `resource_data_dynamic`, I think `dynamic` was only in contrast to `static`. I assume dynamic predicates were so named because their references are stored in `resource_data_dynamic` 

> does each resource have exactly 1 resource logic? 

In Taiga, yes, because all resources need to look identical. But predicates can "call" other predicates. So a "resource logic" could (maybe) consist of 1 "main" predicate and 1 or more "subpredicates", e.g. for efficiency. The resource logic can also "call" some "dynamic" predicates, but we might not want to include those in the "resource logic"

> What is the difference between a "static Predicate" and a "resource logic"? Why have 2 words for what appear to be the same thing?

I agree they are more or less the same thing, unless we choose to distinguish the "main" predicate as "the" resource logic.

**Comment by joebebel** [comment]
>    By the time I read [resource_data_dynamic](https://github.com/anoma/specs/blob/degregat/resource-dag-refactor/src/architecture/layers/resources.md?rgh-link-date=2023-04-29T21%3A56%3A23Z#resource_data_dynamic), I still have no idea what "relevant to the fungibility of a Resource" means. Is the idea here that nothing in resource_data_dynamic should be readable by predicates determining whether this Resource can be created or destroyed? What's the point of this field then? Also, that seems unlikely, given that it would mean the resource logic cannot constrain controllers for the resource, or access "quantity," which seems important.

I think the narrative is confusing because it's an abstraction of concrete things that are influenced a lot by context, and without the context just seem bizarre. 

Here's my understanding of resources. The important preface is that Taiga is reconciling two "competing" execution models, which I'll call "resource balance" vs "resource transition".

In the "resource transition" model, we think of transactions as consuming "state" from a resource (which is subsequently not consumable again) and producing new "state" in a new resource, according to the rules of the resource logic. In this model, there are no other constraints, and resources can be consumed or created arbitrarily. Transaction validity is determined primarily by resource logic.

This is fairly flexible inside a partial transaction, but not at all flexible outside of it, because predicates are constrained in what they can "see". So any transaction which does not fit in a single partial transaction, or is built by multiple agents, is a problem. For example, a simple two party swap isn't so easy, because the state transition is valid only with both Alice and Bob's parts of it, so whoever is the prover needs all of the details from both Alice's and Bob's view.

In the "resource balancing" model, we have some concept of a transaction moving a single resource (or some quantity of fungible resources) from the input to the output. Tokens fall into this model very naturally, because when we consume a token resource and produce a new token resource of the same amount, it's not an arbitrary state transition, but "merely" mutating some data which is changed every time. Transaction validity is determined primarily by _balance_.

In this model, for example, a simple two party swap is not a problem at all because Alice merely spends token A and sends herself token B, which can balance if paired with Bob spending token B and sending himself token A, but Alice and Bob don't have a coordination problem where one of them needs to prove the whole transaction. Indeed, we solved it by *not* allowing arbitrary state transitions across parts of transactions, but enforcing a very particular invariant instead.

So the compromise is that Taiga is a (perhaps unholy) marriage of these two models, which is why it may seem occasionally redundant. In Taiga, a transaction is both a general state transition and a transfer of resources: for a given resource, when this input resource/consumed, some quantity (represented by its logic and static data) is "released" from the input resource and added to the net balance. Similarly, every time this resource is output, that quantity is subtracted from the net balance. So in this way, quantities of resources are preserved.

But Taiga also uses the "resource transition" model: every (partial) transaction consumes state from the input resources, in the form of the input`resource_data_dynamic`, and produces new state in the output resources, in the form of output `resource_data_dynamic`. In this way, Taiga predicates still validate arbitrary state transitions within a partial transaction, while across partial transactions only the balance invariant is preserved. It turns out that this is extremely flexible, because predicates can rely solely on this balance invariant to do complex computations (such as intents) despite the fact that each predicate only gets a "local" view.

**Comment by degregat** [comment]
I just pushed a rewrite of the sections where I think we have rough consensus already, taking into account the feedback from above.

Major open discussion points which remain as far as I can tell are:
- How are commitments and nullifiers handled @joebebel 
- Controllers @isheff @cwgoes 
- How does post ordering execution work @isheff 
- Where do executables come from @isheff 

If everything, apart from the open questions, is roughly correct, I would like to merge this and then continue the discussions, so we don't get too many diverging branches.

**Comment by degregat** [comment]
I'm also wondering if the explanations from above should go into the same document, maybe into an "Intuition" section.

**Comment by degregat** [comment]
> Would you mind rebasing and retargeting this on #122? I think it will be easier.

I'll wait for you to merge and then see if a rebase or a simple merge are easier.

File: src/architecture/layers/resources.md [hunk 1]
```diff
@@ -56,105 +102,130 @@ This way we gain the following options by using signatures of upstream Controlle
...
+## Commitments and Nullifiers
+When a shielded transaction gets executed, the commitments of the resources (equivalent to the Resource Address) created by it are recorded by the consensus provider.
+A verifiable encryption scheme should be used to prove to the consensus provider correspondence between the commitment submitted and an encrypted resource with the new owner as recepient. This encrypted resource can be stored by the consensus provider to guarantee that the recepient can fetch it. 
+
+Nullifiers can always be derived from the plaintext body of a resource and need to be included in the `ptx` consuming them.
+
+## Transactions (tx)
+Transactions provide the notion of balance for a set of `ptx`s, as well as validity for all their Predicates.
+
+A transaction is _balanced_ if and only if the input and output sets of each Resource Type are of the same size.
```
Comment (cwgoes) [inline src/architecture/layers/resources.md:162]: well, including quantity

File: src/architecture/layers/resources.md [hunk 2]
```diff
@@ -56,105 +102,130 @@ This way we gain the following options by using signatures of upstream Controlle
...
+data Executable = Executable {
+  read_keys :: [TyphonDBKey],
+  write_keys :: [TyphonDBKey],
 }
 ```
 
-## Intent
-Intent is encoded in two ways:
-1. In an unbalanced `ptx`, it is expressed as am (Ephemeral) Resource Logic of an Ephemeral Resource.
+An executable contains the machinery to infer from the `ptx`s what is supposed to be read and written to the Typhon DB.
```
Comment (cwgoes) [inline src/architecture/layers/resources.md:200]: let's write this section in a later PR?

File: src/architecture/layers/resources.md [hunk 3]
```diff
@@ -56,105 +102,130 @@ This way we gain the following options by using signatures of upstream Controlle
...
+### Ownership
 Knowledge of the nullifier key of a Resource (i.e. owning the Resource) is necessary, but not necessarily sufficient to own a resource. Ownership might further be constrained via the Predicate which can call on Predicates in `resource_dynamic_data` or external resources.
 
+### Resource Upgrades
+If we want to upgrade the proof system or other static data of a Resource Type, we instantiate a new Resource that is backwards compatible by supporting `ptx`s which take old Resources as inputs. To prevent downgrades, decisions about ordering of "strength" of proof systems should happen at a higher layer and be checked when instantiating Resources for Upgrades.
+
+### Relationship between Notes and Resources
+
+There is a conceptual 1:1 correspondence between a Resource and a Note.
+TODO: These names/objects should be unified and we need to do a thorough examination of what is left to do to move it into practice.
```
Comment (cwgoes) [inline src/architecture/layers/resources.md:227]: has this been done?

**Pull request merged**